### PR TITLE
MotoROS Driver v1.6.0

### PR DESCRIPTION
**Pull request by ted-miller**
- Add fix for the different possible lengths of ROS_MSG_MOTO_JOINT_TRAJ_PT_FULL_EX.
- Add logic to handle multiple messages (or partial messages) coming in on the socket at the same time.
- Removed redundant check for message size.
- Removed startup delay for non-DX100 controllers.
- Added additional configurations of INFORM jobs.

### CODE REVIEW

**Comment by ted-miller** [comment]
v1.6.0
- Added separate server for I/O (based on branch from gavanderhoorn).
- Updated pendant IONAME.DAT file.
- Updated socket select statement to handle concurrent activity on
multiple servers.
- Cleaned up #include statements.

@gavanderhoorn  I just updated this to include your changes from your io-server branch.  My latest commit shows up in the pr "commits" tab, but doesn't appear here in the conversation.  Is that normal?

**Comment by shaun-edwards** [comment]
>@gavanderhoorn I just updated this to include your changes from your io-server branch. My latest commit shows up in the pr "commits" tab, but doesn't appear here in the conversation. Is that normal?

That's not normal.  Click on the "Commits" tab above to make sure everything you want is included in this PR.

**Comment by ted-miller** [comment]
It's all there in commit 18782a6fa438e6817ac0bd771de7d2ac363c295d
(v1.6.0_review)

**Comment by gavanderhoorn** [comment]
Yeah, I'm not sure what happened here. I see all commits as well, but it doesn't show in the *Conversation*.

**Comment by gavanderhoorn** [comment]
General comment / question (which I'd hoped we could've discussed before you merged my changes into this PR): would we want to keep the IO implementation in `MotionServer.c` around for 1 or 2 versions, just to maintain bw-compatibility with non-ROS clients out there?

We can update the ROS nodes in `motoman_driver`, but afaic `MotoROS` is also being used without the ROS components, so it might perhaps be nice to those users to first deprecate the IO infrastructure in `MotoROS` and then remove it in some future version.

Thoughts @ted-miller ?

File: motoman_driver/MotoPlus/MotionServer.c [hunk 1]
```diff
@@ -463,24 +436,26 @@ int Ros_MotionServer_SimpleMsgProcess(Controller* controller, SimpleMsg* receive
 		ret = Ros_MotionServer_JointTrajPtFullExProcess(controller, receiveMsg, replyMsg);
 		break;
 
+
+//Maintain backward compatibility for users who are sending I/O over motion-server
```
Comment (ted-miller) [inline motoman_driver/MotoPlus/MotionServer.c]: @gavanderhoorn  Here is where I continue to process I/O requests on the motion server.  It just calls the io function from ioserver.c

**Comment by ted-miller** [comment]
I just jumped on this right away because I was already tinkering with it and I need to move on to other things next week.

I didn't remove the command from the motion server.  It will accept these commands on either server to maintain backward compatibility.  I plan to create a wiki page at some point that directs people to use the I/O server, but I don't intend to remove the old functionality.

File: motoman_driver/MotoPlus/MotionServer.c (hunk 1 above)
Comment (gavanderhoorn) [reply motoman_driver/MotoPlus/MotionServer.c]: O, that is very nice. I'd missed that. :+1: 

Are the MotoPlus IO functions thread safe? If `MotionServer` and `IoServer` are run in different threads, would `*_ReadIOBit(..)` (and the others) need to introduce locking/mutexes?

**Comment by gavanderhoorn** [comment]
No problem. It was just something I thought we should discuss before merging a potentially breaking change like this.

And I missed that you'd already taken care of it.

>  I plan to create a wiki page at some point that directs people to use the I/O server, but I don't intend to remove the old functionality.

Ok. As long as it doesn't lead to duplication of (maintenance) effort *and* we clearly document the deprecation of the `MotionServer` IO interface.

File: motoman_driver/MotoPlus/MotionServer.c (hunk 1 above)
Comment (ted-miller) [reply motoman_driver/MotoPlus/MotionServer.c]: The I/O API is thread safe.

**Review by gavanderhoorn — CHANGES_REQUESTED** [review]
we've run into some issues while testing this (on our SIA20F): the TP displays an error after installing the MotoPlus application:

> 8000[0]: Could not save PARAMETERS file

Until those are resolved, this should not be merged.

**Review by gavanderhoorn — APPROVED** [review]
Tested this with our SIA20F (+rail): works for me, so +1.

**Comment by gavanderhoorn** [comment]
Error reported in https://github.com/ros-industrial/motoman/pull/183#pullrequestreview-75447433 was due to having multiple instances of the MotoROS application on the controller (one was invisible due to an 'incompatible filename').

After remedying that, everything worked as expected.

**Comment by gavanderhoorn** [comment]
We should pull this into `indigo-devel` as well.

**Comment by gavanderhoorn** [comment]
@shaun-edwards: there are quite some changes missing from `indigo-devel` ([comparison](https://github.com/ros-industrial/motoman/compare/indigo-devel...kinetic-devel)), but all should be Indigo compatible.

Should we do a sync between `kinetic-devel` and `indigo-devel`?

**Pull request merged**